content/docs/userspace/full-stack/2-types.md

```diff
@@ -0,0 +1,215 @@
+
+1. Add a new journal entry.
+2. Edit an existing journal entry.
+3. Delete an existing journal entry.
+
+We can create a tagged union structure for these actions, like so:
+
+```hoon
++$  action
+  $%  [%add =id =txt]
+      [%edit =id =txt]
+      [%del =id]
+  ==
+```
+
+### 3. Updates
+
+Updates are a little more complicated than our actions. Firstly, our front-end
+needs to be able to retrieve an initial list of journal entries to display. Once
+it has that, it also needs to be notified of any changes. For example, if a new
+entry is added, it needs to know so it can add it to the list it's displaying.
+If an entry gets deleted, it needs to remove it from the list. Etc.
+
+The simplest approach to the initial entries is just a `(list entry)`. Then, for
+the subsequent updates, we could send out the `$action`. Since an `$action` is a
+tagged union, it's simpler to have all updates be a tagged union, so when we get
+to doing mark conversions we can just switch on the head tag. Therefore, we can
+define an `$update` structure like so:
+
+```hoon
++$  update
+  $%  action
+      [%jrnl list=(list entry)]
+  ==
+```
+
+There's one drawback to this structure. Suppose either an agent on a remote ship
+or an instance of the front-end client is subscribed for updates, and the
+network connection is disrupted. In the remote ship case, Gall will only allow
+so many undelivered messages to accumulate in Ames before it automatically kicks
+the unresponsive subscriber. In the front-end case, the subscription will also
+be ended if enough unacknowledged messages accumulate, and additionally the
+client may sometimes need to establish an entirely new connection with the ship,
+discarding existing subscriptions. When this happens, the remote ship or web
+client has no way to know how many (if any) updates they've missed.
+
+The only way to resynchronize their state with ours is to discard their existing
+state, refetch the entire initial state once again, and then resubscribe for
+updates. This might be fine if the state of our agent is small, but it becomes a
+problem if it's very large. For example, if our agent holds tens of thousands of
+chat messages, having to resend them all every time anyone has connectivity
+issues is quite inefficient.
+
+One solution to this is to keep an _update log_. Each update can be tagged with
+the time it occurred, and stored in our agent's state, separately to the
+entries. If an agent or web client needs to resynchronize with our agent, it can
+just request all updates since the last one it received. This approach is used
+by the `%graph-store` agent, for example. Our agent is local-only and doesn't
+have a huge state so it might not be strictly necessary, but we'll use it to
+demonstrate the approach.
+
+We can define a logged update like so, where the `@` is the update timestamp in
+milliseconds since the Unix Epoch:
+
+```hoon
++$  logged  (pair @ action)
++$  update
+  %+  pair  @
```

Comment:
named tuple instead of just `pair` w/ `@`?

Reply:
yeah might have been better. I don't think it's too significant though and it will be a pain to change it